tools/vms-generator/vms-generator.go

```diff
@@ -561,31 +564,90 @@ func main() {
 	genDir := flag.String("generated-vms-dir", "", "")
 	flag.Parse()
 
-	var vms = map[string]interface{}{
-		vmiEphemeral:        getVMIEphemeral(),
-		vmiFlavorSmall:      getVMIFlavorSmall(),
-		vmiSata:             getVMISata(),
-		vmiFedora:           getVMIEphemeralFedora(),
-		vmiNoCloud:          getVMINoCloud(),
-		vmiPVC:              getVMIPvc(),
-		vmiWindows:          getVMIWindows(),
-		vmCirros:            getVMCirros(),
-		vmAlpineMultiPvc:    getVMMultiPvc(),
+	var vms = map[string]*v1.VirtualMachine{
+		vmCirros:         getVMCirros(),
+		vmAlpineMultiPvc: getVMMultiPvc(),
+	}
+
+	var vmis = map[string]*v1.VirtualMachineInstance{
+		vmiEphemeral:   getVMIEphemeral(),
+		vmiFlavorSmall: getVMIFlavorSmall(),
+		vmiSata:        getVMISata(),
+		vmiFedora:      getVMIEphemeralFedora(),
+		vmiNoCloud:     getVMINoCloud(),
+		vmiPVC:         getVMIPvc(),
+		vmiWindows:     getVMIWindows(),
+	}
+
+	var vmireplicasets = map[string]*v1.VirtualMachineInstanceReplicaSet{
 		vmiReplicaSetCirros: getVMIReplicaSetCirros(),
-		vmiPresetSmall:      getVMIPresetSmall(),
-		vmTemplateFedora:    getTemplateFedora(),
-		vmTemplateRHEL7:     getTemplateRHEL7(),
-		vmTemplateWindows:   getTemplateWindows(),
 	}
-	for name, obj := range vms {
+
+	var vmipresets = map[string]*v1.VirtualMachineInstancePreset{
+		vmiPresetSmall: getVMIPresetSmall(),
+	}
+
+	var templates = map[string]*Template{
+		vmTemplateFedora:  getTemplateFedora(),
+		vmTemplateRHEL7:   getTemplateRHEL7(),
+		vmTemplateWindows: getTemplateWindows(),
+	}
+
+	handleError := func(err error) {
+		if err != nil {
+			fmt.Printf("%s\n", err)
+			os.Exit(1)
+		}
+	}
+
+	handleCauses := func(causes []metav1.StatusCause, name string, objType string) {
```

Comment:
How is it better though to spill a trace and not a proper error?

Reply:
What we could do instead is actually logging the causes here.

Reply:
In general panic is a very cheap way to see pretty clear where and in which context you fail as an engineer in the code, without the need of thinking too much about what you have to log. I would just have panicked with the validation-reasons/errors.

Reply:
Let me understand if this is a request for change before merge, or just a side note on how you would approach the task. I am fine adopting if that's a request, otherwise I prefer to have clear error messages from the tool.

Note that we may have multiple causes of admission failure, so I would still need to construct a single message to panic with. With the latest upload, we log each cause per line, so that it's clear to the reader what happened.

Reply:
@booxter just replace os.exit(1) with panic(). We'll get the error messages and the stack trace.

Reply:
Yeah as long as we have messages I am good with panic(). I will update tomorrow.

Reply:
It was just a suggestion. Something like validate and panic with the v1.Status data. Works for me like it is right now too.